---
title: "Killifish 16S rRNA Analysis"
author: "Lei Ma"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: TRUE
      smooth_scroll: TRUE
    number_sections: true
    theme: cosmo
---

```{r options, echo=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE, error=FALSE, include=TRUE, fig.width=12, cache=TRUE, collapse=TRUE, autodep=TRUE, fig.path="html_figures/", tidy=TRUE)
knitr::knit_hooks$set(inline = function(x) if(is.numeric(x)){format(x,digits=2)}else{x})
set.seed(1)
```

```{r libraries, cache=FALSE}
#misc
library(here)

#data manipulations
suppressPackageStartupMessages(library(tidyverse))
library(magrittr)
library(readr)
library(broom)

#plotting stuff
library(ggplot2)
library(cowplot)
library(patchwork)
library(kableExtra)

#analysis
suppressPackageStartupMessages(library(dada2))
library(phyloseq)
library(speedyseq)
library(decontam)
library(skimr)
library(vegan)
library(VennDiagram)
library(ape)
library(corncob)
library(cluster)
library(corrplot)

source("../code/helpful_functions.R")
```

# Raw reads and preprocessing

```{r raw data, cache=FALSE}
taxonomy <- read.delim(here("output/taxonomy.txt"), "\t", header=TRUE, row.names=1, check.names = FALSE)
asvs <- read.delim(here("output/ASVs.txt"), "\t", header=TRUE, row.names=1, check.names = FALSE)
metadata <- read.delim(here("data/KLF_metadata.csv"), ",", header = TRUE, row.names=1, check.names=FALSE)
tree <- read_tree(here("output/tree.nwk"))
track <- read.delim(here("output/track.tsv"), "\t", header=TRUE, row.names=1, check.names=FALSE)
track <- mutate(track, prop.retained = nochim/reads.in) 

seqtab <- readRDS(here("output/seqtab_nochimeras.rds"))
rownames(seqtab) <- rownames(track)

mock_ref <- getSequences(here("data/HMP_MOCK.v35.fasta"))

ps <- phyloseq(otu_table(as.matrix(asvs), taxa_are_rows=TRUE),
              sample_data(as.data.frame(metadata)),
              tax_table(as.matrix(taxonomy)),
              phy_tree(tree))

ps <- merge_phyloseq(ps, sample_data(track))
metadata <- as(sample_data(ps), 'data.frame')
```

## Raw data / run summary

Making sure that the two sequencing runs are not too different from each other. In general it seems like Run2 had a higher proportion of reads pass filter but not more overall reads
```{r pass filter}
metadata %>% ggplot(aes(x=nochim, y=prop.retained, color=sequencingRun))+geom_point()+labs(x="Reads pass filter", y="Proportion pass filter")
```
## Mock community check
Next we check that the Mock community is well covered in both runs. Mock1 is from Run1 and Mock2 is from Run2. Looks like there's very low contamination of other sequences in the mock samples. However, there is some slight difference in the ordering of mock abundances, but this variation is similar to previous runs. 

```{r mock}
match_mock <- sapply(names(seqtab[1,]), function(x) any(grepl(x, c(mock_ref))))
match_mock <- rownames(taxonomy[match_mock,])
ps_mock <- ps %>% subset_samples(sampleType=="mock") %>% prune_taxa(taxa_sums(.)>0, .) %>% transform_sample_counts(function(x) x/sum(x)) %>% psmelt()
ps_mock <- ps_mock %>% bind_cols(correct=sapply(ps_mock$OTU, function(x) if(x %in% match_mock){return("Correct")}else{return("Wrong")}))

ps_mock %>% select(OTU, sample, Abundance, correct) %>% group_by(sample, correct) %>% summarise(proportion = sum(Abundance)) %>% pivot_wider(id_cols=sample, values_from = proportion, names_from=correct)

ps_mock %>% filter(OTU %in% match_mock) %>% ggplot(aes(x=reorder(OTU, Abundance), y=Abundance, color=Genus)) + geom_point() + facet_wrap(~sample) + theme(axis.text.x = element_text(angle=45)) + labs(x="ASV", y="Relative abundance")
```

Now that mock is checked out, we can remove it from the data.
```{r remove mock}
ps <- ps %>% subset_samples(sampleType != "mock") %>% prune_taxa(taxa_sums(.)>0,.)
```

## Removing contaminant ASVs with decontam

We will use decontam and the negative controls (both extraction negatives and PCR negatives) to filter out potential contaminant sequences by frequency of appearance in the negative controls.

```{r decontam}
sample_data(ps)$is.neg <- grepl("control", sample_data(ps)$sampleType)
contamdf_prev <- isContaminant(ps, method="prevalence", neg="is.neg")
table(contamdf_prev$contaminant)
```

Below is the taxonomy of the ASVs that were identified as contaminants.

```{r decontam taxonomy}
contaminant_asvs <- rownames(contamdf_prev[which(contamdf_prev$contaminant),])
taxonomy[contaminant_asvs,] %>% kable("html") %>% kable_styling(bootstrap_options = "striped")
```

Here's a graph of the ASV prevalence in true vs negative control samples. Highlighted in blue are the ones identified as contaminants. 
```{r decontam graph}
ps.pa <- transform_sample_counts(ps, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$is.neg == TRUE, ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$is.neg != TRUE, ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf_prev$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")
```

## Chloroplasts

I also want to remove the chloroplasts from the samples. However, I want to make a record of the chloroplast abundance in the metadata of the sample, because it may be useful information later. 

```{r chloroplast levels}
ps.chloroplast <- ps %>% transform_sample_counts(function(x) x/sum(x)) %>% subset_taxa(Order=="Chloroplast") %>% psmelt() %>% select(OTU, sample, Abundance)
ps.chloroplast <- ps.chloroplast %>% group_by(sample) %>% summarise(chl_abundance=sum(Abundance)) %>% column_to_rownames("sample")

ps <- merge_phyloseq(ps, sample_data(ps.chloroplast))

ggplot(sample_data(ps), aes(x=reorder(sample, chl_abundance), y=chl_abundance, color=sampleType)) + geom_point() + facet_wrap(~fishType, scales="free_x") + labs(title="Proportion of reads matching Order Chloroplast", x="Sample", y="Chloroplast relative abundance")
```
Removing chloroplasts, contaminants, and non-bacterial/archaeal ASVs
How do the read counts look after we remove the chloroplast samples and the putative contaminants? 

```{r cleaning up}
# remove contaminants
ps <- prune_taxa(!taxa_names(ps) %in% contaminant_asvs, ps)

# remove chloroplasts and Eukaryotess
ps <- subset_taxa(ps, Kingdom!="Eukaryota") %>% subset_taxa(Order!="Chloroplast" | is.na(Order))

# remove controls
ps <- subset_samples(ps, sampleType %in% c("gut", "water")) %>% prune_taxa(taxa_sums(.)>0,.)

# get number and proportion of reads now
sample_data(ps)$reads.clean <- sample_sums(ps)
sample_data(ps)$final.prop <- sample_data(ps)$reads.clean/sample_data(ps)$reads.in
```

## Cleaned up stats

In the table below nochim is the number of raw reads that passed dada2 filtering. prop.retained is the proportion that passed dada2 filtering. reads.clean is the number of reads we currently have and final.prop is the final proportion of reads pass all the filtering (dada2 and what we just did)
```{r read stats}
read_stats <- sample_data(ps) %>% as_tibble() %>% select(sample, nochim, prop.retained, reads.clean, final.prop) %>% skim(where(is.numeric))
tibble(read_stats) %>% kbl("html") %>% kable_styling(bootstrap_options = "striped")
```

The script preprocessing.R will generate the cleaned data that will be used in the subsequent analysis. After this, I will load the preprocessed .rds files


```{r load processed data, cache=FALSE}
# source(here("scripts/preprocessing.R"))
ps <- readRDS(here("data/processed/ps.rds"))
ps.rel <- readRDS(here("data/processed/psrel.rds"))
ps.melted <- readRDS(here("data/processed/psmelted.rds"))
```


# Diversity analysis

## Alpha diversity

Comparing wild and F2 killifish alpha diversity. Black dots represent the water at each site at the time of sampling. Water for the captive killifish is not available. 

```{r alpha, warning=FALSE}
p <- ps %>% subset_samples(sampleType=="gut") %>% plot_richness(x = "site", measures=c("Observed","Shannon","Simpson"))
w <- ps %>% subset_samples(sampleType=="water") %>% plot_richness(x = "site", measures=c("Observed","Shannon","Simpson"))
ggplot(p$data, aes(x=site, y=value, color=wild_or_F2)) + geom_boxplot() + geom_point(data=w$data, aes(x=site, y=value), color="black") + facet_wrap(~variable, scales="free_y") + theme_light() + labs(title="Alpha diversity metrics", y="")
```
### Alpha diversity significance tests

Pairwise t-tests show that NBH wild fish have significantly different diversity values than all other fish types while SC wild, SC F2, and NBH F2 fish have similar measurements. One exception is that SC wild has a higher Simpson measurement than NBH F2. 
```{r}
rich_est <- estimate_richness(ps, measures=c("Observed", "Shannon","Simpson"))

#make into a tibble with relevant columns from metadata
rich_est <- rich_est %>% rownames_to_column(var="sample") %>% left_join(as(sample_data(ps), "data.frame"), by="sample") %>% filter(sampleType=="gut") %>%  select(sample, Observed, Shannon, Simpson, site, fishType) %>% pivot_longer(cols = Observed:Simpson, names_to = "measurement", values_to = "values")

# do pairwise t tests on each alpha diversity measurement type
rich_est_test <- rich_est %>% group_by(measurement) %>% nest() %>% mutate(test = map(.x=data, ~pairwise.t.test(.x$values, .x$fishType) %>% tidy())) %>% select(measurement, test) %>% unnest(cols=c(test)) %>% mutate(significant = p.value < 0.05) %>% ungroup()
```


## Beta dispersion

Subsetting to just the gut samples, now we look at the beta dispersion/diversity between sites and between F2 and wild type

```{r beta dispersion}
ps.rel.gut <- ps.rel %>% subset_samples(sampleType=="gut")
ps.rel.gut.dist <- distance(ps.rel.gut, "bray")
dat <- as(sample_data(ps.rel.gut), "data.frame")

fishType <- dat$fishType
site <- dat$site
wild_or_F2 <- dat$wild_or_F2

# Check the beta dispersion of site X wild_or_F2
disper.fishType <- betadisper(d=ps.rel.gut.dist, group=fishType)
df <- data.frame(disper=disper.fishType$distances,group=disper.fishType$group)
disper_plot<-ggplot(df, aes(x=group, y=disper)) + geom_boxplot(aes(color=group)) + 
  labs(y="Distance from spatial median (dispersion)") +coord_flip() + scale_x_discrete(limits=rev) + 
  theme_light() + theme(axis.title.y.left = element_blank(), legend.position = "none")
disper_plot
```

Tukey's HSD test shows that all pairwise comparisons of beta dispersion between fish types are significantly different except for between the F2 fish. 
```{r tukey HSD}
tuk.fishType <- TukeyHSD(disper.fishType)
tuk.fishType
```

# Ordination 

## PCoA of all samples

I attempted to do an NMDS ordination but `metaMDS` would not converge. So I created a PCoA using Bray-Curtis dissimilarity
```{r pcoa ordinate}
ps.rel.bc <- distance(ps.rel, method="bray")
ps.rel.bc.pcoa <- ordinate(ps.rel, method="MDS", distance=ps.rel.bc)
```

```{r pcoa plot}
p <- plot_ordination(ps.rel, ps.rel.bc.pcoa) + 
  geom_point(aes(color=site, shape=wild_or_F2), size=4) +
  labs(title="PCoA using Bray Curtis Dissimilarity") +
  theme_light()

# coordinates are fixed to the sqrt of the second eigenvalue divided by the first eigenvalue
p + 
  geom_point(data=p$data %>% filter(sampleType=="water"), aes(x=Axis.1, y=Axis.2, color=site), shape="circle", size=4) + 
  scale_color_manual(name="Site", values=c("pink", "blue")) +
  scale_shape_manual(name="Sample/Fish Type", values=c("square", "circle", "triangle"),labels=c("F2", "Water", "Wild")) +
  coord_fixed(sqrt(ps.rel.bc.pcoa$values[,1][2]/ps.rel.bc.pcoa$values[,1][1]))
```

## Ordination of just wild

```{r}
ps_bc_nmds <- ordinate(subset_samples(ps.rel, wild_or_F2=="wild" & sampleType=="gut") %>% prune_taxa(taxa_sums(.)>0,.), distance="bray", "NMDS", trymax=500)
plot_ordination(subset_samples(ps.rel, wild_or_F2=="wild" & sampleType=="gut") %>% prune_taxa(taxa_sums(.)>0,.), ps_bc_nmds) + geom_point(aes(color=site)) + labs(main="NMDS wild fish")
```

## PERMANOVA of wild and F2 fish

Interestingly a PERMANOVA finds that the two F2 populations are significantly different
```{r}
ps.wild.rel <- subset_samples(ps.rel, wild_or_F2=="wild")
ps.f2.rel <- subset_samples(ps.rel, wild_or_F2=="F2")

adonis2(distance(ps.wild.rel, "bray")~ site + weight.g + length.cm, data=as(sample_data(ps.wild.rel), "data.frame"), by="margin")
adonis2(distance(ps.f2.rel, "bray")~ site + weight.g + length.cm + sex, data=as(sample_data(ps.f2.rel), "data.frame"), by="margin")
```

## PCA Biplot

```{r biplot all}
top10 <- prune_taxa(names(sort(taxa_sums(ps.rel), TRUE))[1:9], ps.rel) %>% otu_table()

biplot.pcoa(x=ps.rel.bc.pcoa, t(top10))
```

```{r pcoa wild}
ps.wild <- ps.rel %>% subset_samples(sampleType=="gut" & wild_or_F2=="wild") %>% prune_taxa(taxa_sums(.)>0,.)
ps.wild.bc <-  ps.wild %>% distance("bray")
ps.wild.pcoa <- ordinate(ps.wild, method="MDS", distance=ps.wild.bc)

p <- plot_ordination(ps.wild, ps.wild.pcoa) + 
  geom_point(aes(color=site), size=4) +
  labs(title="PCoA using Bray Curtis Dissimilarity") +
  theme_light()

# coordinates are fixed to the sqrt of the second eigenvalue divided by the first eigenvalue
p + 
  scale_color_manual(name="Site", values=c("pink", "blue"))
  coord_fixed(sqrt(ps.wild.pcoa$values[,1][2]/ps.wild.pcoa$values[,1][1]))
```
within the wild samples, ASV2, ASV3, and ASV5 structure the populations

```{r}
library(factoextra)
top10 <- prune_taxa(names(sort(taxa_sums(ps.wild), TRUE))[1:9], ps.wild) %>% otu_table()
biplot.pcoa(x=ps.wild.pcoa, t(top10))

```

# ASV differences

## Overlap of ASVs

I was curious how many ASVs were unique to each environment/fish type: Scorton Creek wild, New Bedford Harbor wild, and the respective F2 fish

```{r venn}
taxa_list <- list(nb_w = subset_samples(ps, fishType=="New Bedford Harbor wild") %>% prune_taxa(taxa_sums(.)>0,.) %>% taxa_names(),
                  nb_f2 = subset_samples(ps, fishType=="New Bedford Harbor F2") %>% prune_taxa(taxa_sums(.)>0,.) %>% taxa_names(),
                  sc_w = subset_samples(ps, fishType=="Scorton Creek wild") %>% prune_taxa(taxa_sums(.)>0,.) %>% taxa_names(),
                  sc_f2 = subset_samples(ps, fishType=="Scorton Creek F2") %>% prune_taxa(taxa_sums(.)>0,.) %>% taxa_names())

taxa_overlap <- calculate.overlap(taxa_list)

venn.diagram(taxa_list, filename = here("figures/Venn_gut_color.png"), disable.logging = TRUE, imagetype="png", category.names = c("Tolerant wild", "Tolerant F2", "Sensitive wild", "Sensitive F2"), fill=cols.fishType[1:4], alpha=0.5, cat.col=cols.fishType[1:4], cat.cex=1.5, margin=.05, print.mode="percent")
```

![Venn Diagram](../figures/Venn_gut_color.png)

```{r overlap percents}
f2 <- calculate.overlap(list(taxa_list$sc_f2, taxa_list$nb_f2))

w <- calculate.overlap(list(taxa_list$sc_w, taxa_list$nb_w))
```

F2 fish share `r 100*length(f2$a3)/length(unique(c(f2$a1, f2$a2)))`% of their ASVs while wild fish share `r 100*length(w$a3)/length(unique(c(w$a1, w$a2)))`% of their ASVs.

## Overlap of wild fish with water

```{r venn water}
nb_w <- calculate.overlap(list(nb=subset_samples(ps, fishType=="New Bedford Harbor wild") %>% prune_zeros() %>% taxa_names(),
                               nbw=subset_samples(ps, fishType=="New Bedford Harbor water") %>% prune_zeros() %>% taxa_names()))

sc_w <- calculate.overlap(list(sc=subset_samples(ps, fishType=="New Bedford Harbor wild") %>% prune_zeros() %>% taxa_names(),
                               scw=subset_samples(ps, fishType=="New Bedford Harbor water") %>% prune_zeros() %>% taxa_names()))
```

NBH fish share `r 100*length(nb_w$a3)/length(unique(c(nb_w$a1, nb_w$a2)))`

## Core taxa - present in all fish types

Core taxa in wild type are either Vibrionaceae or Mycoplasma while core taxa in F2 are Lactobacillales
```{r core microbiome}
ps.core <- prune_taxa(taxa_names(ps.rel) %in% taxa_overlap$a6, ps.rel) %>% subset_samples(sampleType=="gut")
core.plot <- ps.core %>% plot_bar(fill="Order")+ geom_bar(aes(fill=Order), stat="identity", position="stack")+facet_wrap(~fishType, scales="free_x")
core.plot.legend <- get_legend(core.plot+ guides(fill = guide_legend(nrow = 3))+theme(legend.position = "bottom"))
p <- core.plot + theme(legend.position = "none")
plot_grid(p, core.plot.legend, ncol=1, rel_heights = c(1,.2))
```

*Core Lactobacillales ASVs:*

```{r}
ps.core %>% tax_table() %>% as.data.frame() %>% filter(Order=="Lactobacillales") %>% .[,4:7] %>% kbl(format="html")
```


*Core Vibrionales ASVs:*
```{r}
ps.core %>% tax_table() %>% as.data.frame() %>% filter(Order=="Vibrionales") %>% .[,4:7] %>% kbl("html")
```


*Core Mycoplasmatales ASVs:*
```{r}
ps.core %>% tax_table() %>% as.data.frame() %>% filter(Order=="Mycoplasmatales") %>% .[4:7] %>% kbl("html")
```


There are still some of these major groups that are not part of the core that are super abundant in some samples. The following bar chart includes all ASVs from the three main orders, not just the core ASVs. Notice the Lactobacillales that appear in a few of the NBH wild fish and the additional Vibrionales in the SC wild fish. 

```{r vibrio myco and lacto}
ps.subset <- subset_taxa(ps.rel, Order %in% c("Vibrionales", "Lactobacillales","Mycoplasmatales")) %>% subset_samples(sampleType=="gut")
ps.subset %>% plot_bar(fill="Order")+ geom_bar(aes(fill=Order), stat="identity", position="stack")+facet_wrap(~fishType, scales="free_x")

```

### What are the top most abundant orders in killifish guts?

I wanted to know what were the most dominant orders in each fish type. So I found the top 5 average relative abundance of order for each fish type and plotted their abundances. 15 Orders had high average relative abundance in at least one fish type, but some were clearly due to a couple of outlier fish.

```{r fifteen top orders}
# in psmelt, group by Order and fishType and summarise average relative abundance, taking the top 5 from each fish type
top5_order <- ps.melted %>% filter(sampleType=="gut") %>% group_by(Order, fishType) %>% summarise(mean_relAbund = mean(abund_rel)) %>% group_by(fishType) %>% top_n(5, mean_relAbund) %>% pull(Order)

# glom ps by order and transform to relative abundance, then melting again
ps_top5order_melt <- ps %>% tax_glom("Order") %>% transform_sample_counts(function(x) x/sum(x)) %>% psmelt() %>% filter(Order %in% top5_order & sampleType=="gut")

# plotting all 15 top Orders
ps_top5order_melt %>% ggplot(aes(x=fishType, y=Abundance)) + geom_boxplot(outlier.shape = NA) + geom_jitter(aes(color=fishType), height = 0, width=.2) + facet_wrap(~Order, scales="free")

```

I removed those orders in which only a few fish showed high relative abundance. Now the number of orders is down to 5 and it's much easier to interpret. Wild fish are dominated by Vibrio and Mycoplasmatales while F2 fish are dominated by Lactobacillales and somewhat by Clostridiales and Cyanobacteriales. I found it interesting that SC wild fish have huge amounts to Vibrio but also huge amounts of Myco, while NBH wild just has a variable amount of vibrio and a small amount of myco. I wonder if the Myco and Vibrio in SC are negatively correlated. For more, see the vibrio section.

```{r five top orders}
# better plot with the clear outlier orders removed
ps_top5order_melt %>% filter(Order %in% c("Vibrionales", "Mycoplasmatales", "Lactobacillales", "Clostridiales", "Cyanobacteriales")) %>% ggplot(aes(x=fishType, y=Abundance)) + geom_boxplot(outlier.shape = NA) + geom_jitter(aes(color=fishType), height = 0, width=.2) + facet_wrap(~Order, scales="free")
```

## Prevalence and abundance

Prevalence abundance analysis shows that the core taxa size is probably less than 100

```{r, include=TRUE}
library(microbiome)
ps.rel.gut <- ps.rel %>% subset_samples(sampleType=="gut") %>% prune_zeros()

p <-plot_core(ps.rel.gut, 
          prevalences = seq(1, 100, 5), 
          detections = seq(0.001, 0.05, length.out=20), 
          plot.type = "lineplot") + 
  xlab("Relative Abundance (%)")
p
```

```{r}
library(RColorBrewer)
p_F2 <- plot_core(ps.rel.gut %>% aggregate_taxa("Taxonomy") %>% subset_samples(wild_or_F2=="F2"),
               prevalences = seq(1,100,5),
               detections=round(.5*10^seq(log10(1e-2), log10(.2), length = 10), 3),
               colours = rev(brewer.pal(5, "RdBu")),
               min.prevalence = 0.1,
               plot.type = "heatmap")
p_wil <- plot_core(ps.rel.gut %>% aggregate_taxa("Taxonomy") %>% subset_samples(wild_or_F2=="wild"),
               prevalences = seq(1,100,5),
               detections=round(.5*10^seq(log10(1e-2), log10(.2), length = 10), 3),
               colours = rev(brewer.pal(5, "RdBu")),
               min.prevalence = 0.1,
               plot.type = "heatmap")
```

```{r}
p_F2 + labs(title="F2 fish core taxa", subtitle = "Minimum prevalence 10%")
p_wil + labs(title="wild fish core taxa", subtitle = "Minimum prevalence 10%")
```

```{r}
library(DirichletMultinomial)
core_wild <- core(ps %>% subset_samples(wild_or_F2=="wild" & sampleType=="gut"), detection = 0.1/100, prevalence = .10)
dat <- abundances(core_wild)
count <- as.matrix(t(dat))
fit_wild <- lapply(1:3, dmn, count = count, verbose=TRUE)
lplc <- base::sapply(fit_wild, DirichletMultinomial::laplace)
best <- fit_wild[[which.min(unlist(lplc))]]
ass <- mixturewt(best) %>% apply(1, which.max)

fitted(best) %>% as.data.frame() %>% rownames_to_column("OTU") %>% set_colnames(c("OTU", "cluster", "values"))

library(reshape2)
for (k in seq(ncol(fitted(best)))) {
  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d <- subset(d, cluster == k) %>%
     # Arrange OTUs by assignment strength
     arrange(value) %>%
     mutate(OTU = factor(OTU, levels = unique(OTU))) %>%
     # Only show the most important drivers
     filter(abs(value) > quantile(abs(value), 0.8))     

  p <- ggplot(d %>% top_n(20, value), aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", k))
  print(p)
}
```

```{r}
core_all <- core(ps %>% subset_samples(sampleType=="gut"), detection = 0.1/100, prevalence = 0.05)
dat <- abundances(core_all)
count <- as.matrix(t(dat))
fit_all <- lapply(1:3, dmn, count = count, verbose=TRUE)
lplc <- base::sapply(fit_all, DirichletMultinomial::laplace)
best <- fit[[which.min(unlist(lplc))]]
ass <- mixturewt(best) %>% apply(1, which.max)

for (k in seq(ncol(fitted(best)))) {
  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d <- subset(d, cluster == k) %>%
     # Arrange OTUs by assignment strength
     arrange(value) %>%
     mutate(OTU = factor(OTU, levels = unique(OTU))) %>%
     # Only show the most important drivers
     filter(abs(value) > quantile(abs(value), 0.8))     

  p <- ggplot(d %>% top_n(20, value), aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", k))
  print(p)
}

```

## Differentially abundant ASVs

**Differentially abundant ASVs between wild and F2**
```{r corncob wild f2, cache=TRUE}
# filter by gut only, and remove ASVs with fewer than 20 reads
# first remove ASVs that never have more than 20 reads
ps.abund <- ps %>% subset_samples(sampleType=="gut") %>% prune_taxa(taxa_sums(.)>20, .)
da_wild_f2 <- differentialTest(formula = ~wild_or_F2,
                               phi.formula = ~wild_or_F2,
                               formula_null = ~1,
                               phi.formula_null = ~wild_or_F2,
                               test="Wald", boot=FALSE,
                               data=ps.abund,
                               fdr_cutoff = 0.05)

# save the test for later
saveRDS(da_wild_f2, file=here("data/processed/da_wild_f2.rds"))
# reading in the .rds
da_wild_f2 <-readRDS(file=here("data/processed/da_wild_f2.rds"))
da_wild_f2_clean <- cleanDA(da_wild_f2, "wild_or_F2")

p <- ggplot(da_wild_f2_clean %>% filter(measure=="mu" & p < 0.05), aes(x=Estimate, y=reorder(str_c(ASV, " ", Taxonomy), Estimate, decreasing=TRUE))) + 
  geom_vline(xintercept = 0, color = "gray50", lty = "dashed", alpha = 0.75, lwd = 1) +
  geom_point() +
  geom_errorbarh(aes(xmin = Estimate-1.96*StdE, xmax = Estimate+1.96*StdE), height = .3) +
  theme_light() +
  facet_wrap(~wild_or_F2, scales = "free_x", nrow = 1) +
  labs(title = "", x = "", y = "Taxa") +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 5)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p + labs(title="Differentially enriched taxa in wild vs F2 fish")
```

```{r da_wild_f2_bar_plot}
ps_sub <- metadata %>% filter(sampleType=="gut") %>% group_by(wild_or_F2) %>% sample_n(10) %>% use_series(sample)
ps_sub <- ps %>% subset_samples(sample %in% ps_sub) %>% tax_glom(taxrank="Family") %>% transform_sample_counts(~ ./sum(.)) %>% psmelt()

sig_family <- da_wild_f2_clean$Family %>% unique() %>% .[1:5]

ps_sub <- ps_sub %>% mutate(wild_or_F2 = factor(wild_or_F2)) %>% mutate(Family_f = factor(Family, levels=c(sig_family, ps_sub$Family %>% unique() %>% .[!. %in% sig_family])))

labels_family <- c(levels(ps_sub$Family_f)[1:5], "Others",
             paste0("Others", 1:(length(unique(ps_sub$Family_f)) - 6)))

labels_family <- c(levels(ps_sub$Family_f)[1:5], rep("Others", times=length(unique(ps_sub$Family_f)) - 5))
ps_sub <- ps_sub %>%
  mutate(Family_f = factor(Family_f, labels = labels_family))

cols_family <- c(hcl.colors(5, palette = "viridis"), rep("gray", length(ps_sub$Family_f %>% unique)-5))

p <- ggplot(ps_sub, aes(x=sample, y = Abundance, fill=Family_f)) + 
  geom_bar(stat="identity") + 
  facet_wrap(~ wild_or_F2, scales="free_x") + 
  scale_fill_manual("Family", 
                    breaks = c(levels(ps_sub$Family_f)[1:5], "Others"),
                    values = cols_family) + 
  ylab("Relative Abundance") + 
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) + 
  theme_minimal() + 
  theme(axis.text.x = element_blank())
```

**Differentially abundant taxa between F2 SC and NBH fish**
```{r corncob NBH SC f2, cache=TRUE}
ps.f2.abund <- ps %>% subset_samples(wild_or_F2=="F2" & sampleType=="gut") %>% prune_taxa(taxa_sums(.)>20, .)
da_site_f2 <- differentialTest(formula = ~site,
                               phi.formula = ~site,
                               formula_null = ~1,
                               phi.formula_null = ~site,
                               test="Wald", boot=FALSE,
                               data=ps.f2.abund,
                               fdr_cutoff = 0.05)
da_site_f2_clean <- cleanDA(da_site_f2, "site")

saveRDS(da_site_f2, file=here("data/processed/da_site_f2.rds"))


r <- ggplot(da_site_f2_clean %>% filter(measure=="mu" & p < 0.05), aes(x=Estimate, y=reorder(str_c(ASV, " ", Taxonomy), Estimate, decreasing=TRUE))) + 
  geom_vline(xintercept = 0, color = "gray50", lty = "dashed", alpha = 0.75, lwd = 1) +
  geom_point() +
  geom_errorbarh(aes(xmin = Estimate-1.96*StdE, xmax = Estimate+1.96*StdE), height = .3) +
  theme_light() +
  facet_wrap(~site, scales = "free_x", nrow = 1) +
  labs(title = "", x = "", y = "Taxa") +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 5)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
r + labs(title="Differentially enriched taxa between F2 SC and wild NBH fish")
```


**Differentially abundant ASVs between NBH wild and SC wild**
```{r corncob NBH SC wild, cache=TRUE}
ps.wild.abund <- ps %>% subset_samples(wild_or_F2=="wild" & sampleType=="gut") %>% prune_taxa(taxa_sums(.)>20, .)
da_site_wild <- differentialTest(formula = ~site,
                               phi.formula = ~site,
                               formula_null = ~1,
                               phi.formula_null = ~site,
                               test="Wald", boot=FALSE,
                               data=ps.wild.abund,
                               fdr_cutoff = 0.05)

#saveRDS(da_site_wild, file=here("data/processed/da_site_wild.rds"))
da_site_wild <- readRDS(file=here("data/processed/da_site_wild.rds"))
da_site_wild_clean <- cleanDA(da_site_wild, "site")

q <- ggplot(da_site_wild_clean %>% filter(measure=="mu" & p < 0.05), aes(x=Estimate, y=reorder(str_c(ASV, " ", Taxonomy), Estimate, decreasing=TRUE))) + 
  geom_vline(xintercept = 0, color = "gray50", lty = "dashed", alpha = 0.75, lwd = 1) +
  geom_point() +
  geom_errorbarh(aes(xmin = Estimate-1.96*StdE, xmax = Estimate+1.96*StdE), height = .3) +
  theme_light() +
  facet_wrap(~site, scales = "free_x", nrow = 1) +
  labs(title = "", x = "", y = "Taxa") +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 5)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
q + labs(title="Differentially enriched taxa between wild SC and wild NBH fish")
```


### Abundance of significant taxa in the seawater

Are these taxa just enriched in the seawater of the respective sites? Only a small fraction of the significant ASVs are more abundant in SC water if they were enriched in the guts or less abundant in SC water if they were depleted in guts. 
```{r sig water}
sig_water <- ps.rel %>% prune_taxa(taxa_names(ps.rel) %in% da_site_wild_clean$ASV,.) %>% subset_samples(sampleType=="water") %>% psmelt() %>% rename(ASV="OTU") %>% left_join(da_site_wild_clean[,1:6], by="ASV")
ggplot(sig_water, aes(x=Abundance, y=reorder(str_c(ASV, " ", Taxonomy), Estimate, decreasing=TRUE), shape=site)) +
  scale_y_discrete(limits=rev) +
  geom_point(aes(color=ifelse(Estimate>0, "red", "green"))) + scale_x_log10() +
  labs(y="signifiant taxa", title="Abundance of significant taxa in SW") +
  scale_color_manual(labels=c("no", "yes"), values=c("red", "blue")) + guides(color=guide_legend(title="enriched in SC"))
```

```{r sig water table}
sig_water <- sig_water %>% select(ASV, site, Estimate, Taxonomy, Abundance, Sample) %>% mutate(SC_enriched = Estimate >0) %>% group_by(ASV, site) %>% summarise(avg=mean(Abundance), SC_enriched = SC_enriched, Taxonomy=Taxonomy) %>% distinct() %>% pivot_wider(id_cols=c(ASV,site, SC_enriched, Taxonomy), names_from=site, values_from=avg) %>% mutate(expected = ifelse(SC_enriched, ifelse(`Scorton Creek` > `New Bedford Harbor`, "yes", "no"),ifelse(`New Bedford Harbor` > `Scorton Creek`, "yes", "no")))
sig_water <- rename(sig_water, NBH_abund="New Bedford Harbor", SC_abund = "Scorton Creek")
sig_water[order(sig_water$expected),] %>% kbl("html") %>% kable_styling(bootstrap_options = "striped")
```

# Vibrionales

In the core taxa section, I found that Vibrionales is highly abundant in wild fish, and particularly in Scorton Creek wild fish. In the differential abundance analysis, I found that it's a specific vibrio ASV (ASV2) that is enriched in SC fish (both wild and F2) while NBH wild fish are enriched in different vibrio taxa. I think this difference warrants further investigation.

## Vibrionales differences between the wild fish

I looked at the overlap in vibrio ASVs between the two fish populations. 

```{r}
ps_vibrio <- ps %>% subset_samples(sampleType=="gut" & wild_or_F2=="wild") %>% subset_taxa(Order=="Vibrionales") %>% prune_taxa(taxa_sums(.)>0,.)

taxa_list <- list(
  SC = subset_samples(ps_vibrio, fishType=="Scorton Creek wild") %>% prune_taxa(taxa_sums(.)>0,.) %>% taxa_names(),
  NBH = subset_samples(ps_vibrio, fishType=="New Bedford Harbor wild") %>% prune_taxa(taxa_sums(.)>0,.) %>%  taxa_names())

overlap <- calculate.overlap(taxa_list)
```

Scorton Creek wild fish have `r length(overlap$a1)` Vibrio ASVs, New Bedford wild fish have `r length(overlap$a2)` Vibrio ASVs, and they share `r length(overlap$a3)` Vibrio ASVs. For clarity, I only plotted the ASVs which ever reach a relative abundance above 1%. In SC ASV2 is dominant, but in NBH ASV2 and ASV5 are about equally dominant

```{r vibrio overlap}
ps_vibrio_overlap <- ps_vibrio %>% filter_taxa2(~ sum(.>0)>3) %>% merge_samples2(group="site") %>%  transform_sample_counts(~ ./sum(.)) %>% psmelt()


# ps_vibrio_overlap <- ps %>% subset_samples(wild_or_F2 == "wild" | sampleType=="water") %>% subset_taxa(Order=="Vibrionales") %>% filter_taxa2(~ sum(.>0)>3) %>% merge_samples2(group="fishType") %>% transform_sample_counts(~ ./sum(.)) %>% psmelt()

sig_vibrios <- da_site_wild_clean %>% filter(Order=="Vibrionales" & measure=="mu") %>% use_series("ASV")
cols.vibrio <- c(hcl.colors(length(sig_vibrios), palette="viridis"), gray.colors(44-length(sig_vibrios)))
p <- ps_vibrio_overlap %>% ggplot(aes(x="", y=Abundance, fill=OTU)) + geom_bar(width=1, stat="identity") + coord_polar("y", start=0) + facet_wrap(~site) + theme_void() + scale_fill_manual(values=cols.vibrio, breaks=c(sig_vibrios, p$data$OTU %>% unique() %>% .[!. %in% sig_vibrios]))

# p <- ps_vibrio_overlap %>% ggplot(aes(x="", y=Abundance, fill=OTU)) + geom_bar(width=1, stat="identity") + coord_polar("y", start=0) + facet_wrap(~fishType) + theme_void() + scale_fill_manual(values=cols.vibrio, breaks=c(sig_vibrios, p$data$OTU %>% unique() %>% .[!. %in% sig_vibrios]))

p
```

```{r, eval=FALSE}
ps_vibrio_overlap <- ps_vibrio %>% transform_sample_counts(~ ./sum(.)) %>% filter_taxa(function(x) any(x>0.01), prune=TRUE) %>% psmelt() %>% filter(OTU %in% overlap$a3)
shading <- tibble(OTU = unique(ps_vibrio_overlap$OTU)) %>% mutate(min=seq(from=.5, to=max(as.numeric(as.factor(OTU))), by=1), max=min+1, col=rep(c(0,1), length.out=length(min)))
p <- ps_vibrio_overlap %>% ggplot() + geom_boxplot(aes(y=Abundance, x=reorder(OTU, -Abundance), color=site)) + geom_rect(data=shading, aes(xmin=min, xmax=max, ymin=-Inf, ymax=Inf, fill=factor(col)), alpha=0.2) + scale_fill_manual(values=c("white", "gray53")) + theme_bw() + guides(fill="none") + theme(legend.direction = "horizontal", legend.position = "bottom") + labs(x="", y="Relative abundance", main="Vibrio ASVs in wild fish")
```

ASV2 is way more abundant in SC wild and somewhat more abundant in SC F2 (though lower)

```{r}
ps.rel %>% subset_samples(sampleType=="gut") %>% prune_taxa("ASV2", .) %>% psmelt() %>% ggplot(aes(x=fishType, y=Abundance, color=fishType)) +geom_boxplot() + geom_point() + facet_wrap(~wild_or_F2, scales="free") + scale_color_manual(values=cols.fishType[1:4]) + theme_cowplot() + theme(axis.text.x=element_blank(), legend.position = "none") + labs(x="", y="Relative abundance")
```

## Tree of Vibrios

```{r vibrio tree}

library(ggtree)

vibrio_tree <- phy_tree(tree) %>% merge_phyloseq(ps_vibrio) %>% prune_taxa(overlap$a3,.) %>%  phy_tree() %>% ggtree(ladderize = F) + geom_tiplab(aes(label=label, color=label))
vibrio_tree
```
## Vibrio corrplot 

This is a shitty corrplot that doesn't take into account the compositionality of the data, but does show that ASV2 is negatively correlated with many of the other vibrios.

```{r vibrio corrplot}
v_otu <- ps_vibrio %>% transform_sample_counts(function(x) x/sum(x)) %>% filter_taxa(function(x) any(x>0.01), prune=TRUE) %>% otu_table() %>% t() %>% as.data.frame()

v_otu <- ps_vibrio %>% subset_samples(fishType=="New Bedford Harbor wild") %>% transform_sample_counts(function(x) x/sum(x)) %>% filter_taxa(function(x) any(x>0.01), prune=TRUE) %>% otu_table() %>% t() %>% as.data.frame()
corrplot(cor(v_otu[,rev(vibrio_tree$data$label[1:46])]))
```

A corrplot made from the sparse covariance matrix of wild fish using Spieceasi doesn't show ASV2 anticorrelated with other vibrios

```{r vibrio corrplot better}
se_cov <- readRDS(here("data/processed/se_wild_cov.rds"))
ps.wild.abund <- ps %>% subset_samples(sampleType=="gut" & wild_or_F2=="wild") %>% prune_taxa(taxa_sums(.)>50,.)
colnames(se_cov) <- rownames(se_cov) <- rownames(otu_table(ps.wild.abund))

se_cov[ps.wild.abund %>% subset_taxa(taxa_names(ps.wild.abund) %in% overlap$a3) %>% taxa_names(),ps.wild.abund %>% subset_taxa(taxa_names(ps.wild.abund) %in% overlap$a3) %>% taxa_names()] %>% as.matrix() %>% cov2cor() %>% corrplot()
```   

# Mycoplasmatales

Mycoplasma abundance not associated with weight or fish size

```{r}
ps.rel %>% subset_taxa(Order=="Mycoplasmatales") %>% subset_samples(wild_or_F2=="wild") %>% psmelt() %>% group_by(sample) %>% summarise(sum = sum(Abundance), weight.g=mean(weight.g), length.cm=mean(length.cm), site=site) %>% filter(sum>0) %>% ggplot(aes(x=sum, y=weight.g/length.cm)) + geom_point() + facet_wrap(~site) + scale_x_log10()
```

# Simper analysis

What taxa are contributing most to B-C dissimilarities?

```{r}
asvs_gut <- ps  %>% subset_samples(sampleType=="gut") %>% prune_taxa(taxa_sums(.)>50,.)%>% otu_table() %>% as.data.frame()
metadata_gut <- ps %>% subset_samples(sampleType=="gut") %>% sample_data() %>% as.data.frame()

simp_fishType <- simper(t(asvs_gut), metadata_gut$fishType)
simp_summary <- summary(simp_fishType)
simp_summary
```

# Network analysis

```{r}
library(SpiecEasi)
library(igraph)
library(tidygraph)
library(ggraph)
library(graphlayouts)

# load results
load(here("data/processed/se_results.RData"))

# recreate the ps objects that were used as the input for SpiecEasi
ps_NBH_wild <- ps %>% subset_samples(fishType=="New Bedford Harbor wild") %>% prune_taxa(taxa_sums(.)>50,.)
ps_SC_wild <- ps %>% subset_samples(fishType=="Scorton Creek wild") %>% prune_taxa(taxa_sums(.)>50,.)
ps_NBH_F2 <- ps %>% subset_samples(fishType=="New Bedford Harbor F2") %>% prune_zeros()
ps_SC_F2 <- ps %>% subset_samples(fishType== "Scorton Creek F2") %>% prune_zeros()
ps_wild <- ps %>% subset_samples(sampleType=="gut" & wild_or_F2=="wild") %>% prune_taxa(taxa_sums(.)>50,.)

se_wild_icov <- readRDS(here("data/processed/se_wild_icov.rds"))

taxonomy <- tax_table(ps) %>% as.data.frame() %>% rownames_to_column(var="ASV")

# Make network/igraph objects
se_SC_net <- adj2igraph(se_SC_refit, vertex.attr = list(name=taxa_names(ps_SC_wild)))
se_NBH_net <- adj2igraph(se_NBH_refit %>% symBeta(), vertex.attr = list(name=taxa_names(ps_NBH_wild)))
se_SC_F2_net <- adj2igraph(se_SC_F2_refit, vertex.attr = list(name=taxa_names(ps_SC_F2)))
se_NBH_F2_net <- adj2igraph(se_NBH_F2_refit, vertex.attr = list(name=taxa_names(ps_NBH_F2)))

# name cov matrix dimensions
colnames(se_NBH_cov) <- rownames(se_NBH_cov) <- rownames(otu_table(ps_NBH_wild))
colnames(se_SC_cov) <- rownames(se_SC_cov) <- rownames(otu_table(ps_SC_wild))
colnames(se_wild_icov) <- rownames(se_wild_icov) <- rownames(otu_table(ps_wild))

# Add attributes to the nodes
V(se_SC_net)$degree <- degree(se_SC_net)
V(se_NBH_net)$degree <- degree(se_NBH_net)
V(se_SC_F2_net)$degree <- degree(se_SC_F2_net)
V(se_NBH_F2_net)$degree <- degree(se_NBH_F2_net)

V(se_SC_net)$relabund <- apply(ps.rel %>% subset_taxa(taxa_names(ps.rel) %in% taxa_names(ps_SC_wild)) %>% otu_table(), 1, function(x) log(mean(x)))
V(se_NBH_net)$relabund <-apply(ps.rel %>% subset_taxa(taxa_names(ps.rel) %in% taxa_names(ps_SC_wild)) %>% otu_table(), 1, function(x) log(mean(x)))
V(se_SC_F2_net)$relabund <- apply(ps.rel %>% subset_taxa(taxa_names(ps.rel) %in% taxa_names(ps_SC_F2)) %>% otu_table(), 1, function(x) log(mean(x)))
V(se_NBH_F2_net)$relabund <-apply(ps.rel %>% subset_taxa(taxa_names(ps.rel) %in% taxa_names(ps_NBH_F2)) %>% otu_table(), 1, function(x) log(mean(x)))

# Use tidygraph to add more node attributes
se_SC_tidynet <- as_tbl_graph(se_SC_net) %>% activate(nodes) %>% mutate(trans_local = local_transitivity(), degree_local = local_ave_degree()) %>% 
  left_join(taxonomy, by=c("name"="ASV"))

se_NBH_tidynet <- as_tbl_graph(se_NBH_net) %>% activate(nodes) %>% mutate(trans_local = local_transitivity(), degree_local = local_ave_degree()) %>% 
  left_join(taxonomy, by=c("name"="ASV"))

se_SC_F2_tidynet <- as_tbl_graph(se_SC_F2_net) %>% activate(nodes) %>% mutate(trans_local = local_transitivity(), degree_local = local_ave_degree()) %>% 
  left_join(taxonomy, by=c("name"="ASV"))

se_NBH_F2_tidynet <- as_tbl_graph(se_NBH_F2_net) %>% activate(nodes) %>% mutate(trans_local = local_transitivity(), degree_local = local_ave_degree()) %>% 
  left_join(taxonomy, by=c("name"="ASV"))
```

## Topological indices for SC and NBH wild networks

Compared to the average degree, the degree distribution is more unequal in the smaller network. So bigger nodes are more popular/important/connected than in the larger network. Smaller network has no bridging nodes between disconnected subgraphs. Shortest path measure as a measure of how well the network is connected, but it excludes missing paths. Centralization: global network measure expressing the degree to which the centrality of the most central node exceeds the centrality of all other nodes in the network. 
```{r}
top_ind <- tribble(
  ~"Topological Indices", ~"NBH wild", ~"SC wild", ~"NBH F2", ~"SC F2",
  "No. ASVs (nodes)", length(V(se_NBH_net)), length(V(se_SC_net)), length(V(se_NBH_F2_net)), length(V(se_SC_F2_net)),
  "No. interactions (edges)", length(E(se_NBH_net)), length(E(se_SC_net)), length(E(se_NBH_F2_net)), length(E(se_SC_F2_net)),
  "Mean degree", round(mean(degree(se_NBH_net)),0), round(mean(degree(se_SC_net)),0), round(mean(degree(se_NBH_F2_net)),0), round(mean(degree(se_SC_F2_net)),0),
  "SD degree", round(sd(degree(se_NBH_net)),0), round(sd(degree(se_SC_net)),0), round(sd(degree(se_NBH_F2_net)),0), round(sd(degree(se_SC_F2_net)),0),
  "Transitivity/Cluster coefficient", transitivity(se_NBH_net), transitivity(se_SC_net), transitivity(se_NBH_F2_net), transitivity(se_SC_F2_net),
  "Avg. path length", mean_distance(se_NBH_net), mean_distance(se_SC_net), mean_distance(se_NBH_F2_net), mean_distance(se_SC_F2_net),
  "Density", edge_density(se_NBH_net), edge_density(se_SC_net), edge_density(se_NBH_F2_net), edge_density(se_SC_F2_net),
  "Modularity (fast-greedy)", modularity(cluster_fast_greedy(se_NBH_net)), modularity(cluster_fast_greedy(se_SC_net)), modularity(cluster_fast_greedy(se_NBH_F2_net)), modularity(cluster_fast_greedy(se_SC_F2_net)),
  "Degree centralization", centr_degree(se_NBH_net)$centralization, centr_degree(se_SC_net)$centralization, centr_degree(se_NBH_F2_net)$centralization, centr_degree(se_SC_F2_net)$centralization,
  "Number of components, excluding singletons", sum(components(se_NBH_net)$csize>1), sum(components(se_SC_net)$csize>1), sum(components(se_NBH_F2_net)$csize>1), sum(components(se_SC_F2_net)$csize>1))
# number of nodes (ASVs)
# number of interactions (edges)
# mean degree +/- SD
# transitivity aka cluster coefficient
# average path length
# average neighbors
# density 
# heterogeneity

top_ind
```


## NetSwan attack tolerance test

Because there is more community structure, SC is more sensitive to loss of connectivity/loss of bridging nodes. 

```{r}
library(NetSwan)

# Swan combinatory took a long time so these results were run once and then saved

# swan_NBH_wild <- swan_combinatory(se_NBH_net, 10)
# swan_SC_wild <- swan_combinatory(se_SC_net, 10)
# save(swan_NBH_wild, swan_SC_wild, file=here("data/processed/netswan_wild.RData"))

load(here("data/processed/netswan_wild.RData"))
colnames(swan_NBH_wild) <- colnames(swan_SC_wild) <- c("fraction", "betweenness", "degree", "cascade", "random")
swan_wild <- pivot_longer(swan_NBH_wild %>% as_tibble(), cols = betweenness:random, names_to = "attack") %>% add_column(fishType=rep("New Bedford Harbor wild", n=nrow(.))) %>% bind_rows(pivot_longer(swan_SC_wild %>% as_tibble(), cols=betweenness:random, names_to = "attack") %>% add_column(fishType=rep("Scorton Creek wild", n=nrow(.))))
# ggplot(swan_wild, aes(x=fraction, y=value, color=site)) + geom_point() + facet_wrap(~attack)

p<-ggplot(swan_wild %>% filter(attack=="degree"), aes(x=fraction, y=value, color=fishType)) + geom_point(size=4) + scale_color_manual(values = c(cols.fishType[1], cols.fishType[3])) + labs(x="Fraction of nodes removed", y=expression("Loss of connectivity \U2192"))
p + theme(axis.title = element_text(size=20), legend.title = element_blank())
```

## Plotting the networks

More specialized networks in non-NBH-wild fish?
```{r plot network wild}
library(RColorBrewer)
# make color palette of 28 distinct colors
graph_color_phyla <- tibble(breaks=unique(as_tibble(se_NBH_tidynet)$Phylum), values=c("#d25032","#576bd8","#5bc24f","#9459d2","#babb3a","#d265cf","#58952c","#c94097","#6cb66c","#e03d6b","#58c99e","#8d53a5","#958e2c","#a495e1","#dc913b","#536db1","#bdaf6c","#52a3d8","#a84b3d","#42c0c7","#b24669","#3d7f46","#dd87bb","#61712c","#95527d","#348a6e","#e4877b","#956a32"))

# Make layout using igraph's "nicely" algo
sc_layout <- create_layout(se_SC_tidynet, layout="stress")
sc_wild_graph <- ggraph(sc_layout) + geom_edge_link(edge_colour="grey66") + geom_node_point(aes(color=Phylum), size=.3) + scale_color_manual(breaks=graph_color_phyla$breaks, values=graph_color_phyla$values) + labs(title = "Scorton Creek wild") + theme(legend.position = "none")

# remove an outlying node to improve plotting
nbh_layout <- create_layout(se_NBH_tidynet %>% filter(name!="ASV702"), layout="stress")
nbh_wild_graph <- ggraph(nbh_layout) + geom_edge_link(edge_color="grey66", edge_alpha=0.15) + geom_node_point(aes(color=Phylum), size=.3) + scale_color_manual(breaks=graph_color_phyla$breaks, values=graph_color_phyla$values) + labs(title="New Bedford Harbor wild") + theme(legend.position = "none")

sc_F2_layout <- create_layout(se_SC_F2_tidynet, layout="stress")
sc_F2_graph <- ggraph(sc_F2_layout) + geom_edge_link(edge_colour="grey66") + geom_node_point(aes(color=Phylum), size=.3) + scale_color_manual(breaks=graph_color_phyla$breaks, values=graph_color_phyla$values) +labs(title="Scorton Creek F2") + theme(legend.position = "none")

nbh_F2_layout <- create_layout(se_NBH_F2_tidynet, layout="stress")
nbh_F2_graph <- ggraph(nbh_F2_layout) + geom_edge_link(edge_color="grey66", edge_alpha=0.15) + geom_node_point(aes(color=Phylum), size=.3) + scale_color_manual(breaks=graph_color_phyla$breaks, values=graph_color_phyla$values) + labs(title="New Bedford Harbor F2") + theme(legend.position = "none")

all_net <- plot_grid(sc_wild_graph, nbh_wild_graph, sc_F2_graph, nbh_F2_graph)

ggsave(all_net,filename=here("figures/plot_se_graphs.png"), width=5, height=5, units="in")
```

## Key player analysis

```{r}
# install package keyplayer

```

## Degree distributions of NBH and SC wild networks

But maybe this comparison isn't valid because there's such a big difference in node number

```{r degree distribution}
# compute and plot degree distributions
deg_dist <- tibble(fishType="Scorton Creek wild", deg = degree.distribution(se_SC_net)) %>% 
  bind_rows(tibble(fishType="Scorton Creek F2", deg = degree.distribution(se_SC_F2_net))) %>% 
  bind_rows(tibble(fishType="New Bedford Harbor wild", deg = degree.distribution(se_NBH_net))) %>% 
  bind_rows(tibble(fishType="New Bedford Harbor F2", deg = degree.distribution(se_NBH_F2_net))) %>% 
  group_by(fishType) %>% mutate(x = 0:(n()-1))

deg_dist$fishType_f <- factor(deg_dist$fishType, levels = c("Scorton Creek wild", "New Bedford Harbor wild", "Scorton Creek F2", "New Bedford Harbor F2"))
ggplot(deg_dist, aes(x=x, y=deg, color=fishType)) + geom_point(shape=1)+geom_line()+ facet_wrap(~fishType_f, scales="free") + scale_color_manual(values=cols.fishType) +
  labs(x="Degree", y="Frequency") + theme_cowplot()+ theme(legend.position = "none")

```


## Pruning NBH wild to have same # V as SC wild

I pruned NBH taxa such that it had the same number of taxa as SC wild did in the original network and reran the network test. 
```{r}
load(here("data/processed/se_tests_refits.RData"))

ps_sc_pruned <- ps %>% subset_samples(fishType=="Scorton Creek wild") %>% prune_taxa(taxa_sums(.)>50,.)
ps_nbh_top <- ps  %>% subset_samples(fishType=="New Bedford Harbor wild")  %>% prune_taxa(names(sort(taxa_sums(.), TRUE))[1:ntaxa(ps_sc_pruned)],.)


se_NBH_top_net <- adj2igraph(se_test_nbh_top_gl %>% symBeta(), vertex.attr = list(name=taxa_names(ps_nbh_top)))
V(se_NBH_top_net)$degree <- degree(se_NBH_top_net)

taxonomy <- tax_table(ps) %>% as.data.frame() %>% rownames_to_column(var="ASV")


se_NBH_top_tidynet <- as_tbl_graph(se_NBH_top_net) %>% activate(nodes) %>% mutate(trans_local = local_transitivity(), degree_local = local_ave_degree()) %>% 
  left_join(taxonomy, by=c("name"="ASV"))

graph_color_phyla <- tibble(breaks=unique(as_tibble(se_NBH_top_tidynet)$Phylum), 
                            values=c("#d25032","#576bd8","#5bc24f","#9459d2","#babb3a","#d265cf","#58952c","#c94097","#6cb66c","#e03d6b","#58c99e","#8d53a5","#958e2c","#a495e1","#dc913b","#536db1","#bdaf6c","#52a3d8","#a84b3d","#42c0c7"))
nbh_top_layout <- create_layout(se_NBH_top_tidynet, layout="stress")
nbh_wild_top_graph <- ggraph(nbh_top_layout) + geom_edge_link(edge_color="grey66", edge_alpha=0.15) + geom_node_point(aes(color=Phylum), size=.3) + scale_color_manual(breaks=graph_color_phyla$breaks, values=graph_color_phyla$values) + labs(title="New Bedford Harbor wild") + theme(legend.position = "none")

dd.nbh <- degree.distribution(se_NBH_top_net)

plot(0:(length(dd.nbh)-1), dd.nbh, type='b',
     ylab="Frequency", xlab="Degree", main="Degree Distributions")

net_stats_NBH_top <- tribble(
  ~"Topological Indices", ~"NBH wild subset",
  "No. ASVs (nodes)", length(V(se_NBH_top_net)),
  "No. interactions (edges)", length(E(se_NBH_top_net)),
  "Mean degree", round(mean(degree(se_NBH_top_net)),0),
  "SD degree", round(sd(degree(se_NBH_top_net)),0),
  "Transitivity/Cluster coefficient", transitivity(se_NBH_top_net),
  "Avg. path length", mean_distance(se_NBH_top_net),
  "Density", edge_density(se_NBH_top_net),
  "Modularity (fast-greedy)", modularity(cluster_fast_greedy(se_NBH_top_net)),
  "Degree centralization", centr_degree(se_NBH_top_net)$centralization,
  "Number of components, excluding singletons", sum(components(se_NBH_top_net)$csize>1))
left_join(top_ind, net_stats_NBH_top)
```

## Wild microbiomes in general

I found that SC fish had high abundances of both vibrios and mycoplasmatales and wondered if they were negatively correlated overall and if Vibrio and Mycoplasmatales collectively structure wild (or just SC) gut microbiomes. 

```{r}
vibrios <- ps_wild %>% tax_table() %>% as.data.frame() %>% rownames_to_column(var="ASV") %>% filter(Family=="Vibrionaceae")

vibrio_icov<-se_wild_icov[rownames(se_wild_icov) %in% vibrios$ASV,colnames(se_wild_icov) %in% vibrios$ASV]
heatmap(vibrio_icov %>% as.matrix())

vibrio_net <- graph_from_adjacency_matrix(vibrio_icov, weighted=TRUE) %>% as_tbl_graph()
ggraph(vibrio_net) + geom_edge_link(aes(edge_color=weight)) + geom_node_point() + scale_edge_color_gradient2(low="blue", high = "red", mid="white", midpoint = 0)

#turns out all vibrios are negatively covarying with each other
```

# Host AHR genotype

Let's look at the distribution of host genotypes. I think it's interesting that the distribution of genotypes is so similar for the wild fish but different in the F2 that were common garden raised. Fish with genotype "none" did not have a band wither either set of primer. I have re-extracted DNA and re-run the primers for the wild fish, which decreased the number of "none" genotypes, but have not done so yet for the F2 fish, hence the greater number of "none" fish in F2. 

```{r host genotypes}
ggplot(metadata %>% filter(!is.na(genotype) & sampleType=="gut"), aes(x=genotype)) + geom_bar(position="stack") +facet_wrap(~fishType) + labs("Distribution of genotypes before removing high chloroplast fish") + geom_text(stat='count', aes(label=..count..), vjust=1, color="yellow")
```

## Wild fish microbiome correlation with AHR genotype

Not looking great visually
```{r ahr bc}
ps.rel.filtered <- ps.rel %>% subset_samples(wild_or_F2=="wild" & sampleType=="gut" & genotype != "none") %>% prune_taxa(taxa_sums(.)>0,.)
p.ord <- ordinate(ps.rel.filtered, method="NMDS", distance = "bray", trymax=500)
plot_ordination(ps.rel.filtered, p.ord, color="genotype") + labs(title="BC NMDS")
```

Permanova confirms no correlation 
```{r ahr permanova}
p.bc <- distance(ps.rel.filtered, "bray")
genotype <- sample_data(ps.rel.filtered)$genotype

adonis2(p.bc ~ genotype, as(sample_data(ps.rel.filtered),"data.frame"))
```